openedx/core/lib/call_stack_manager.py

```diff
@@ -0,0 +1,111 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+CallStackManager -  stores all stacks in global dictionary and logs
+CallStackMixin - used for Model save(), and delete() method
+
+Functions:
+capture_call_stack - global function used to store call stack
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager, CallStackMixin
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+3. For tracking Save and Delete events-
+    Use mixin called "CallStackMixin"
+    For ex.
+        class StudentModule(CallStackMixin, models.Model):
+
+Note -
+1.Format for STACK_BOOK
+{
+    "modelclass1":
+        [[(frame 1),(frame 2)],
+         [(frame 11),(frame21)]]
+    "modelclass2":
+        [[(frame 3),(frame 4)],
+         [(frame 6),(frame 5)]]
+
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = {}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks.
+EXCLUDE = ['^.*python2.7.*$', '^.*call_stack_manager.*$']
+REGULAR_EXPS = [re.compile(x) for x in EXCLUDE]
+
+
+def capture_call_stack(current_model):
+    """
+    stores customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+
+    Arguments:
+    current_model - Name of the model class
+    """
+    # holds temporary callstack
+    temp_call_stack = [(line.split(',')[0].strip().replace("\n", "")[6:-1],
+                        line.split(',')[1].strip().replace("\n", "")[6:],
+                        line.split(',')[2].strip().replace("\n", "")[3:])
```

Comment:
It's not ideal to be splitting each line 3 times like this. It's also a bit unclear what operation is actually being performed (what are you extracting from the line?). It might be clearer if you pulled out the line-munging into a separate function.

Also, what is each line? Using `replace('\n', '')` seems to indicate that it might be many lines, not just one. Maybe it's really a frame?